tunelo/api/utils.py

```diff
@@ -12,3 +18,198 @@ def make_error_response(message=None, status_code=None):
+    """Resolves a channel's type as determined by its ``device_owner`` attribute
+    as such:
+
+    ``"channel:<channel_type>:(spoke|hub)"``
+
+    Raises:
+        MalformedChannel: If this channel has an invalid ``device_owner``
+    """
+    device_owner = get_channel_device_owner(port)
+    valid_device_owner = device_owner_pattern.match(device_owner)
+    if not valid_device_owner:
+        raise MalformedChannel(
+            f"Port {port['id']} has invalid device_owner: {device_owner}."
+        )
+    channel_type = valid_device_owner.group("channel_type")
+    if channel_type not in VALID_CHANNEL_TYPES:
+        raise MalformedChannel(f"Channel type '{channel_type}' is not supported.")
+    return channel_type
+
+
+def get_channel_network_id(port):
+    return port["network_id"]
+
+
+def get_channel_device_owner(port):
+    return port["device_owner"]
+
+
+get_spoke_channel_public_key = partial(get_binding_profile_attribute, attr="public_key")
+
+
+def get_channel_uuid(port):
+    return port["id"]
+
+
+def get_channel_endpoint(port):
+    return port["endpoint"]
+
+
+def get_channel_status(port):
+    return port["status"]
+
+
+def get_channel_project_id(port):
+    return port["project_id"]
+
+
+def get_channel_peers_spokes(spokes, hubs):
+    """Gets a peers for a collection of spoke ports
+
+    A peer for a spoke channel is, for now, defined as any hub channel which has
+    a spoke's public key in its peers list.
+
+    This function fetches peers for multiple spokes, which makes it more efficient
+    for use in ListChannel.
+
+    Args:
+        spokes: An iterable of spoke port definitions, which are generated by
+        ``neutron_client.get_ports()``
+        hubs: An iterable of hub port definitions, which are generated by
+        ``neutron_client.get_ports()``
+
+    Raises:
+        MalformedChannel: If a hub port has an invalid peer entry
+
+    Returns:
+        A dictionary, mapping spoke UUIDs to lists of hubs which have them as peers
+    """
+    if not hubs:
+        return {get_channel_uuid(s): [] for s in spokes}
+    if not spokes:
+        return {}
+
+    # Maps spoke public keys to hubs which have a peer with a given spoke public key
+    hubs_by_peer_representation = {}
+
+    # We determine a hub to be a peer if it has an entry with ``port``'s public key
+    # in its own list of peers
+    for hub in hubs:
+        peers = get_channel_properties(hub).get("peers")
+        if not peers:
+            continue
+        for peer in peers:
+            # Hub ports have peer entries which represent spokes.
+            # These entries look like "pubkey|endpoint|allowed_ips
+            valid_peer = valid_hub_peer_pattern.match(peer)
+            if not valid_peer:
+                raise MalformedChannel(
+                    f"Hub channel {get_channel_uuid(hub)} "
+                    f"has invalid peer entry: {peer}"
+                )
+            hubs_by_peer_representation.setdefault(peer, [])
```

Comment:
Given that this is how you're using it, you can make `hubs_by_peer_representation` a `defaultdict(list)` instead.